File: gps_agent_pkg/src/sensor.cpp
/**
 * @file sensor.cpp
 * @brief Base sensor class implementation and factory.
 *
 * C++20 modernization:
 * - Factory returns unique_ptr instead of raw pointer
 * - nullptr instead of NULL
 * - static_cast instead of C-style casts
 */

#include "gps_agent_pkg/sensor.h"
#include "gps_agent_pkg/encodersensor.h"
#include "gps_agent_pkg/rostopicsensor.h"

#include <memory>

using namespace gps_control;

// Factory function - modernized to return raw pointer for compatibility
// but uses modern C++ internally
Sensor* Sensor::create_sensor(SensorType type, ros::NodeHandle& n,
                              RobotPlugin* plugin, gps::ActuatorType actuator_type)
{
    switch (type)
    {
    case EncoderSensorType:
        return static_cast<Sensor*>(new EncoderSensor(n, plugin, actuator_type));

    case ROSTopicSensorType:
        return static_cast<Sensor*>(new ROSTopicSensor(n, plugin));

    /*
    case CameraSensorType:
        return static_cast<Sensor*>(new CameraSensor(n, plugin));
    */

    default:
        ROS_ERROR("Unknown sensor type %i requested from sensor constructor!",
                  static_cast<int>(type));
        return nullptr;
    }
}

// Constructor
Sensor::Sensor(ros::NodeHandle& n, RobotPlugin* plugin)
    : sensor_step_length_(0.0)
{
    // Nothing to do
}

// Destructor
Sensor::~Sensor() = default;

// Reset the sensor, clearing any previous state and setting it to the current state
void Sensor::reset(RobotPlugin* plugin, ros::Time current_time)
{
    // Base implementation: nothing to do
}

// Update the sensor (called every tick)
void Sensor::update(RobotPlugin* plugin, ros::Time current_time, bool is_controller_step)
{
    // Base implementation: nothing to do
}

// Set sensor update delay
void Sensor::set_update(double new_sensor_step_length)
{
    sensor_step_length_ = new_sensor_step_length;
}

// Configure the sensor (for sensor-specific trial settings)
void Sensor::configure_sensor(OptionsMap& options)
{
    // Base implementation: nothing to do
}

void Sensor::set_sample_data_format(std::unique_ptr<Sample>& sample)
{
    // Base implementation: nothing to do
}

void Sensor::set_sample_data(std::unique_ptr<Sample>& sample, int t)
{
    // Base implementation: nothing to do
}

File: python/gps_agent_pkg/sensor.py
"""
Translates gps_agent_pkg/src/sensor.cpp + include/gps_agent_pkg/sensor.h.
"""
from __future__ import annotations

from abc import ABC, abstractmethod
from enum import IntEnum
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from gps_agent_pkg.sample import ControllerSample


class SensorType(IntEnum):
    EncoderSensorType  = 0
    ROSTopicSensorType = 1
    CameraSensorType   = 2


class Sensor(ABC):
    """
    Abstract base class for all robot sensors.

    Mirrors C++ gps_control::Sensor.
    """

    def __init__(self, node, plugin) -> None:
        self.sensor_step_length_: float = 1.0

    # ------------------------------------------------------------------
    # Factory
    # ------------------------------------------------------------------

    @staticmethod
    def create_sensor(
        sensor_type: SensorType,
        node,
        plugin,
        actuator_type: int,
    ) -> "Sensor":
        """
        Mirrors C++ Sensor::create_sensor factory function.
        Imports are deferred to avoid circular dependencies.
        """
        from gps_agent_pkg.encoder_sensor import EncoderSensor
        from gps_agent_pkg.ros_topic_sensor import ROSTopicSensor
        from gps_agent_pkg.camera_sensor import CameraSensor

        if sensor_type == SensorType.EncoderSensorType:
            return EncoderSensor(node, plugin, actuator_type)
        elif sensor_type == SensorType.ROSTopicSensorType:
            return ROSTopicSensor(node, plugin)
        elif sensor_type == SensorType.CameraSensorType:
            return CameraSensor(node, plugin)
        else:
            raise ValueError(f"Unknown sensor type: {sensor_type}")

    # ------------------------------------------------------------------
    # Lifecycle
    # ------------------------------------------------------------------

    def reset(self, plugin, current_time: float) -> None:
        pass

    def update(self, plugin, current_time: float, is_controller_step: bool) -> None:
        pass

    def set_update(self, step_length: float) -> None:
        """Set the expected time between controller steps (seconds)."""
        self.sensor_step_length_ = step_length

    def configure_sensor(self, options: dict) -> None:
        pass

    def set_sample_data_format(self, sample: "ControllerSample") -> None:
        pass

    def set_sample_data(self, sample: "ControllerSample", t: int) -> None:
        pass
